### ISSUE TITLE
GPU and CPU gradients diverge in TF 2.4 for approximate gelu activation

### ISSUE THREAD

**Issue by benjrossi** [tensorflow/tensorflow#47724]
Hi,

Upon upgrading from TF 2.2 to TF 2.4, my team noticed a problem with an approximate version of `gelu` shown below as `gelu_approximate`, which exists also in TF v2.4 as `tf.nn.gelu(approximate=True)`. The gradients for `gelu_approximate` calculated on the GPU diverge from the gradients on the CPU as the input gets farther from 0. Since the implementation of `gelu_approximate` consists of primitives `tanh`, `pow`, etc, we should make sure something more fundamental isn't broken in TF. Please see the following reproduce code, which prints `"GPU and CPU gradients are not close!"` for v2.4 but not for v2.2.

```python
# gelu_problem.py
import numpy as np
import pandas as pd
import tensorflow as tf
import matplotlib.pyplot as plt
 

def gelu_approximate(x):
    # Copied and inlined from tf.nn.gelu(approximate=True) which exists in TF v2.4 but not TF v2.2
    return 0.5 * x * (1.0 + tf.tanh(0.7978845608028654 * (x + 0.044715 * tf.pow(x, 3))))


def gelu_gradient(x, device):
    with tf.GradientTape() as tape:
        tape.watch(x)
        with tf.device(device):
            y = gelu_approximate(x)
    return tape.gradient(y, x)
 
 
def main():
    print(f"TF version is {tf.__version__}")
    x = tf.linspace(-500.0, 500.0, 500)
    cpu = gelu_gradient(x, "/CPU:0")
    gpu = gelu_gradient(x, "/GPU:0")
    abs_error = np.abs(cpu - gpu)
    df = pd.DataFrame(dict(cpu=cpu, gpu=gpu, abs_error=abs_error), index=x)
    try:
        np.testing.assert_allclose(cpu, gpu, atol=1e-3)
    except (AssertionError,) as e:
        print("GPU and CPU gradients are not close!")
        print(e)
    df.plot(title="CPU vs. GPU gradients for tf.nn.gelu(approximate=True)")
    plt.show()
 
 
if __name__ == "__main__":
    main()
```

Output on our system:

```
$ nvidia-smi
Wed Mar 10 15:49:02 2021      
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 450.80.02    Driver Version: 450.80.02    CUDA Version: 11.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla V100-PCIE...  Off  | 00000000:04:00.0 Off |                    0 |
| N/A   39C    P0    42W / 250W |  29877MiB / 32510MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-PCIE...  Off  | 00000000:06:00.0 Off |                    0 |
| N/A   40C    P0    38W / 250W |  31045MiB / 32510MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-PCIE...  Off  | 00000000:07:00.0 Off |                    0 |
| N/A   40C    P0    37W / 250W |  31045MiB / 32510MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  Tesla V100-PCIE...  Off  | 00000000:08:00.0 Off |                    0 |
| N/A   41C    P0    38W / 250W |  31045MiB / 32510MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                                
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A   4126251      C   .../3/7/x/dist/bin/python3.7    29871MiB |
|    1   N/A  N/A   4126251      C   .../3/7/x/dist/bin/python3.7    31039MiB |
|    2   N/A  N/A   4126251      C   .../3/7/x/dist/bin/python3.7    31039MiB |
|    3   N/A  N/A   4126251      C   .../3/7/x/dist/bin/python3.7    31039MiB |
+-----------------------------------------------------------------------------+
 
$ python -um gelu_problem
2021-03-10 15:48:44.908579: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.10.1
TF version is 2.4.0
2021-03-10 15:48:50.149409: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-03-10 15:48:50.150776: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1
2021-03-10 15:48:50.220817: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties:
pciBusID: 0000:04:00.0 name: Tesla V100-PCIE-32GB computeCapability: 7.0
coreClock: 1.38GHz coreCount: 80 deviceMemorySize: 31.75GiB deviceMemoryBandwidth: 836.37GiB/s
2021-03-10 15:48:50.220842: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.10.1
2021-03-10 15:48:50.224004: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.10
2021-03-10 15:48:50.224038: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublasLt.so.10
2021-03-10 15:48:50.226493: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcufft.so.10
2021-03-10 15:48:50.227454: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcurand.so.10
2021-03-10 15:48:50.229775: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusolver.so.10
2021-03-10 15:48:50.231379: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusparse.so.10
2021-03-10 15:48:50.235484: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.7
2021-03-10 15:48:50.237107: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1862] Adding visible gpu devices: 0
2021-03-10 15:48:50.237501: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE3 SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-03-10 15:48:50.239017: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-03-10 15:48:50.239769: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties:
pciBusID: 0000:04:00.0 name: Tesla V100-PCIE-32GB computeCapability: 7.0
coreClock: 1.38GHz coreCount: 80 deviceMemorySize: 31.75GiB deviceMemoryBandwidth: 836.37GiB/s
2021-03-10 15:48:50.239789: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.10.1
2021-03-10 15:48:50.239805: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.10
2021-03-10 15:48:50.239819: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublasLt.so.10
2021-03-10 15:48:50.239832: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcufft.so.10
2021-03-10 15:48:50.239845: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcurand.so.10
2021-03-10 15:48:50.239858: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusolver.so.10
2021-03-10 15:48:50.239872: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusparse.so.10
2021-03-10 15:48:50.239885: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.7
2021-03-10 15:48:50.241280: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1862] Adding visible gpu devices: 0
2021-03-10 15:48:50.241307: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.10.1
2021-03-10 15:48:51.004798: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1261] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-03-10 15:48:51.004834: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1267]      0
2021-03-10 15:48:51.004839: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 0:   N
2021-03-10 15:48:51.010315: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 1908 MB memory) -> physical GPU (device: 0, name: Tesla V100-PCIE-32GB, pci bus id: 0000:04:00.0, compute capability: 7.0)
GPU and CPU gradients are not close!
 
Not equal to tolerance rtol=1e-07, atol=0.001
 
Mismatched elements: 466 / 500 (93.2%)
Max absolute difference: 3.1899037
Max relative difference: 1.
 x: array([ 0.      ,  0.      ,  0.      ,  0.      ,  0.      ,  0.      ,
        0.      ,  0.      ,  0.      ,  0.      ,  0.      ,  0.      ,
        0.      ,  0.      ,  0.      ,  0.      ,  0.      ,  0.      ,...
 y: array([-3.189903e+00, -3.151703e+00, -3.113807e+00, -3.076217e+00,
       -3.038931e+00, -3.001947e+00, -2.965265e+00, -2.928882e+00,
       -2.892799e+00, -2.857012e+00, -2.821522e+00, -2.786328e+00,...
```

![gelu_problem](https://user-images.githubusercontent.com/18267168/110800398-21220300-824a-11eb-85a9-4dfee7af8334.png)

This particular system is running Debian Linux, version 9.13. Python is version 3.7.9, TF is version 2.4.0. CUDNN is 7.6.5 and CUDA is 10.1.

Thank you,

Ben and team

**Comment by reedwm** [comment]
`tf.nn.gelu` doesn't exist in TF 2.2, as it was introduced in TF 2.4. How were you able to run it in TF 2.2?

**Comment by benjrossi** [comment]
> `tf.nn.gelu` doesn't exist in TF 2.2, as it was introduced in TF 2.4. How were you able to run it in TF 2.2?

@reedwm our mistake when editing this example to make it shorter. (However, the original issue still applies). We had copied the function as it exists in TF 2.4 into our own function. I had edited it out just before this submission believing that `tf.nn.gelu` was in both versions, to make the submission shorter. Please just inline and copy the contents of `tf.nn.gelu` when the `approximate=True` branch taken, instead of calling `tf.nn.gelu`, to get a bug-free repro in TF v2.2. This is what we had done initially. I can update the example above to reflect that change. Thanks.

**Comment by benjrossi** [comment]
For more context, the approximation we used prior to TF 2.4 came from [this paper](https://arxiv.org/pdf/1606.08415.pdf). The `tf.nn.gelu(approximate=True)` happens to boil down to the same function we were using before the upgrade.

**Comment by reedwm** [comment]
Thank you for filing this issue. If tanh is run on the CPU, then your example passes, so the isuse is with tanh:

```python
# gelu_problem.py, but tanh is run on CPU which causes this to work correctly
import numpy as np
import pandas as pd
import tensorflow as tf
import matplotlib.pyplot as plt


def tanh_cpu(x):
  with tf.device('/CPU:0'):
    return tf.tanh(x)


def gelu_approximate(x, run_tanh_on_cpu):
  tanh = tanh_cpu if run_tanh_on_cpu else tf.tanh
  # Copied and inlined from tf.nn.gelu(approximate=True) which exists in TF v2.4 but not TF v2.2
  return 0.5 * x * (1.0 + tanh(0.7978845608028654 * (x + 0.044715 * tf.pow(x, 3))))


def gelu_gradient(x, device, run_tanh_on_cpu):
  with tf.GradientTape() as tape:
    tape.watch(x)
    with tf.device(device):
      y = gelu_approximate(x, run_tanh_on_cpu)
  return tape.gradient(y, x)


def main():
  print(f"TF version is {tf.__version__}")
  x = tf.linspace(-500.0, 500.0, 500)
  cpu = gelu_gradient(x, "/CPU:0", run_tanh_on_cpu=False)
  gpu = gelu_gradient(x, "/GPU:0", run_tanh_on_cpu=False)
  abs_error = np.abs(cpu - gpu)
  df = pd.DataFrame(dict(cpu=cpu, gpu=gpu, abs_error=abs_error), index=x)
  try:
    np.testing.assert_allclose(cpu, gpu, atol=1e-3)
  except (AssertionError,) as e:
    print("GPU and CPU gradients are not close!")
    print(e)
  df.plot(title="CPU vs. GPU gradients for tf.nn.gelu(approximate=True)")
  plt.show()


if __name__ == "__main__":
  main()
``` 

In particular, running tanh on a very large float32 input incorrectly gives 0.99999976 instead of 1.0.

```
import tensorflow as tf
with tf.device('/GPU:0'):
  print(tf.tanh(tf.constant(1000., dtype=tf.float32)))  # 0.99999976, but should be 1.
```

When given very large inputs, tanh should round to 1 as the true result is much closer to 1 than 0.99999976. tanh gives the correct result on float16 or float64 values, just not float32 values. I do not understand gelu and tanh well enough to understand why the slightly incorrectly tanh output causes the gelu gradient to have such a large gradient, but fixing tanh should fix the issue.

I bisected this and the issue first started occurring at c6023a81d4976f6ff79f957925364d21d7884004. @akuegel, can you fix the GPU tanh op when given large float32 inputs?

**Comment by sherhut** [comment]
I have verified that with a recent build (after the above commit)

```
import tensorflow as tf
with tf.device('/GPU:0'):
  print(tf.tanh(tf.constant(1000., dtype=tf.float32)))
```

produces 1.0 as desired and your example also now works. Closing as fixed.

**Comment by benjrossi** [comment]
Thank you, @sherhut.

**Comment by oawxkw** [comment]
I also observed the following API aliases or similar APIs can cause the same issue in some older versions of tensorflow.
Users should be cautious when using them on GPU from tensorflow 2.4.0 (v2.4.0-rc4-71-g582c8d236cb) to 2.4.3 (v2.4.2-142-g72bb4c22adb).

> Besides, in tensorflow 2.3.x, these APIs throw the InternalError Exception on GPU, which is also unexpected.

- `tf.tanh`, `tf.math.tanh`, `tf.nn.tanh`, `tf.compat.v1.tanh`, `tf.compat.v1.math.tanh`, `tf.compat.v1.nn.tanh`
- `tf.keras.activations.tanh`, `tf.compat.v1.keras.activations.tanh`
- `tf.experimental.numpy.tanh`

It seems to be fixed in tensorflow 2.5.0 (v2.5.0-rc3-213-ga4dfb8d1a71) and later versions.
